### PR TITLE
Inline smaller callees first

### LINKED ISSUE #130590: `Range<usize>::next` should fully MIR-inline
If you compile this to optimized MIR:
```rust
// @compile-flags: -O -C debuginfo=0 --emit=mir -Z inline-mir-threshold=9999 -Z inline-mir-hint-threshold=9999
use std::ops::Range;
#[no_mangle]
pub fn demo(num: &mut Range<usize>) -> Option<usize> {
    num.next()
}
```
<https://rust.godbolt.org/z/zsh6b6Y8n>

You'll see that it still contains a call to `forward_unchecked`:
```rust
    bb1: {
        _3 = copy ((*_1).0: usize);
        StorageLive(_4);
        _4 = <usize as Step>::forward_unchecked(copy _3, const 1_usize) -> [return: bb2, unwind continue];
    }
```

That's pretty unfortunate, because `forward_unchecked(a, 1)` is just `AddUnchecked(a, 1)`, a single MIR operator.

### PR DESCRIPTION
**Pull request by scottmcm**
Then limit the total size and number of inlined things, to allow more top-down inlining (particularly important after calling something generic that couldn't inline internally) without getting exponential blowup.

Fixes #130590
r? saethlin

### CODE REVIEW

File: compiler/rustc_mir_transform/src/inline.rs [hunk 1]
```diff
@@ -102,46 +103,80 @@ fn inline<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) -> bool {
...
     };
-    let blocks = START_BLOCK..body.basic_blocks.next_index();
-    this.process_blocks(body, blocks);
-    this.changed
+    let initial_blocks = START_BLOCK..body.basic_blocks.next_index();
+    this.enqueue_new_candidates(body, initial_blocks);
+
+    let mut changed = false;
+    let mut remaining_cost = tcx.sess.opts.unstable_opts.inline_mir_total_threshold.unwrap_or(300);
+    let mut remaining_count = MAX_INLINED_CALLEES_PER_BODY;
```
Comment (the8472) [inline compiler/rustc_mir_transform/src/inline.rs:117]: Shouldn't this go by some size heuristic? if the callee is absolutely tiny (e.g. just another call) then I'd expect the cost of inlining to be ~zero (replacing a call with another call), which means we can inline an unlimited amount of those cases.

Comment (the8472) [reply compiler/rustc_mir_transform/src/inline.rs:117]: Ah, looking at the testcases I guess the scopes make this non-zero cost?

**Comment by saethlin** [comment]
@rust-timer build 018727d766cb382a9e117d9d50a6abc0f5bd7ef0

**Comment by oli-obk** [comment]
We could limit it to nonincremental release builds.

It may also be of interest to embedded folk when optimizing for size

**Comment by oli-obk** [comment]
Considering incremental regressed, I'd say we're indeed running on level 3

**Comment by bjorn3** [comment]
When optimizations are enabled, `-Zmir-opt-level=2` is used by default and `-Zmir-opt-level=1` when optimizations are disabled. Incr check and debug builds also regress. It probably has more to do with this optimization running on the standard library.

**Comment by saethlin** [comment]
> Considering incremental regressed

Any improvement to the inliner tends to regress incremental. I think this happens because the standard library is not built with incremental, so better inlining means more of the standard library gets dragged into third party CGUs.

**Comment by the8472** [comment]
> > Considering incremental regressed
> 
> Any improvement to the inliner tends to regress incremental. I think this happens because the standard library is not built with incremental, so better inlining means more of the standard library gets transitively dragged into third party CGUs.

This is for generic MIR exported from the library or for prebuilt monomorphic code? Would it make sense to distinguish between MIR for export (and maybe defer optimization to the consuming crate?) and MIR generated for same-crate codegen?

**Comment by saethlin** [comment]
> This is for generic MIR exported from the library or for prebuilt monomorphic code?

It's for MIR exported from the library; not all of that MIR is generic.

> and maybe defer optimization to the consuming crate?

I suspect this would have similarly problematic consequences for incremental builds.

What might help is having two version of `optimized_mir`, one which has all interprocedural optimizations disabled and is used for incremental builds.

**Pull request closed without merging**